### PR TITLE
OpenSSL V3 (3.0) support

### PR DESCRIPTION
**Pull request by OverOrion**
This PR adds support for warning-free compilation with OpenSSL 3.0.
Kudos to @MrAnno

A few notes:
- MD4 hash template function will be deprecated  (see 67ae1c3 for details)
- Ubuntu 22.04 ships with OpenSSL 3.0 by default if you want to check the warning-free compilation without compiling OpenSSL for yourself
- Please do not merge it yet, I still want to run some (internal) tests

I was hesitant to move the modifications regarding `cmac` function to the compat layer, my reasoning:
-  It is in the `secure-log` module meaning that it is acceptable to directly interface with the OpenSSL API

If it should be in the compat layer, then please let me know and I will move it accordingly.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/secure-logging/slog.c [hunk 1]
```diff
@@ -420,15 +424,34 @@ gchar *convertToBase64(unsigned char *input, gsize len)
  */
 void cmac(unsigned char *key, const void *input, gsize length, unsigned char *out, gsize *outlen)
 {
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
```
Comment (alltilla) [inline modules/secure-logging/slog.c:427]: Reflecting on
> It is in the secure-log module meaning that it is acceptable to directly interface with the OpenSSL API

Generally it is allowed to use OpenSSL API directly from any module. Just like before `tlscontext: move setup methods of DH-params to openssl-compat`, the `lib/tlscontext.c` file used them. Most of the functions in `lib/compat/openssl_support.c` are redefinitions of original OpenSSL functions for the sake of version compatibility, but sometimes we put "helper" functions there, just like you did in the mentioned commit.

I think what you did in that commit is really similar to moving out this function.

You might ask: what do we gain by moving these helper functions to the compat layer? I think it is better to see the OpenSSL version preprocessor `#if`s in one separate file, than them being everywhere in the codebase. So I would prefer moving these functions there.

File: modules/secure-logging/slog.c [hunk 2]
```diff
@@ -420,15 +424,34 @@ gchar *convertToBase64(unsigned char *input, gsize len)
...
 
   CMAC_Init(ctx, key, KEY_LENGTH, EVP_aes_256_cbc(), NULL);
   CMAC_Update(ctx, input, length);
 
-  size_t outsize;
-  CMAC_Final(ctx, out, &outsize);
-  *outlen = outsize;
+  size_t out_len;
+  CMAC_Final(ctx, out, &out_len);
+  *outlen = out_len;
```
Comment (alltilla) [inline modules/secure-logging/slog.c:452]: [nitpicking, no need to change anything]

Reformatting lines of code, which we did not modify could go to separate commits. Same applies for removing dead code (`sLogDecrypt()` in the header file).

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
OpenSSL 3.0 is already available in debian-sid. Hopefully, it will reach debian-testing soon.

File: lib/compat/openssl_support.c [hunk 3]
```diff
@@ -184,6 +185,78 @@ void openssl_ctx_setup_ecdh(SSL_CTX *ctx)
 #endif
 }
 
+long
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply lib/compat/openssl_support.c]: Yes, gbool would be better for methods that are not strict OpenSSL compat functions.

File: modules/secure-logging/slog.c [hunk 4]
```diff
@@ -420,15 +424,34 @@ gchar *convertToBase64(unsigned char *input, gsize len)
...
+    OSSL_PARAM_utf8_string("cipher", "aes-256-cbc", 0),
+    OSSL_PARAM_END,
+  };
+
+  EVP_MAC_CTX *ctx = EVP_MAC_CTX_new(mac);
+
+  EVP_MAC_init(ctx, key, KEY_LENGTH, params);
+  EVP_MAC_update(ctx, input, length);
+  size_t out_len;
+  EVP_MAC_final(ctx, out, &out_len, 99999);
```
Comment (MrAnno) [inline modules/secure-logging/slog.c]: This was probably an experiment, please fix the 4. parameter.

File: modules/secure-logging/slog.c (hunk 1 above)
Comment (MrAnno) [reply modules/secure-logging/slog.c]: cmac() is not general enough in its current form (and is not well-audited) to be moved under compat/openssl.

We have to be really careful what we put there. IMHO the openssl_support unit should contain trivial copy-pated compatibility functions, other non-trivial helpers, but we should not place custom cryptographic primitives there unless they are general enough to be used anywhere and they are audited properly.

File: modules/secure-logging/slog.c (hunk 2 above)
Comment (MrAnno) [reply modules/secure-logging/slog.c]: Because crypto-related code is modified here, and we almost missed a bug in the few lines above, I think it would be best to split this commit to make it as atomic and clean as possible.

File: lib/compat/openssl_support.c [hunk 5]
```diff
@@ -333,3 +338,38 @@ openssl_ctx_setup_session_tickets(SSL_CTX *ctx)
   SSL_CTX_set_num_tickets(ctx, 0);
 #endif
 }
+
+void
+openssl_cmac(unsigned char *key, size_t key_size, const void *input, gsize length, unsigned char *out, gsize *outlen)
```
Comment (MrAnno) [inline lib/compat/openssl_support.c:343]: I think we should not do this in its current form. See my comment above.

File: modules/secure-logging/slog.c (hunk 4 above)
Comment (OverOrion) [reply modules/secure-logging/slog.c]: Thank you, I completely missed it when I started working again on this branch

**Comment by OverOrion** [comment]
Thank you @alltilla, @MrAnno for your notes.
Based on them:
- `cmac` stays in its original place
- the dummy 4th parameter in  `EVP_MAC_final` is now the correct value*
- `SYSLOG_NG_HAVE_DECL_DIGEST_MD4` macro is defined as `1`

*Because of propagation it was necessary to change the `cmac` and its callers's API that's why there are so many commits.
The change is introduced in 23fc6ee21c851e49094d14a1c3f3da5811a3880f and the commits after it are follow-ups for each function.
Finally, to eliminate the array calculating logic, the `GET_BUF_SIZE` is introduced.

Kindly check them. Thanks in advance!

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
@OverOrion I might have said something confusing in my comment about atomicity.
Could you please squash the commits that follow the new cmac parameter up?
They are hard to revert/bisect/review this way (they don't compile, which is one of our requirements).

File: modules/secure-logging/slog.c [hunk 6]
```diff
@@ -493,19 +493,20 @@ void PRF(unsigned char *key, unsigned char *originalInput, guint64 inputLength,
 
   // Make sure that temporary buffer can hold at least outputLength bytes, rounded up to a multiple of AES_BLOCKSIZE
   unsigned char buf[outputLength+AES_BLOCKSIZE];
+  gsize buf_capacity = sizeof(buf) / sizeof(buf[0]);
   // Prepare plaintext
   for (int i=0; i<outputLength/AES_BLOCKSIZE; i++)
     {
       gsize outlen;
-      cmac(key, input, AES_BLOCKSIZE, &buf[i*AES_BLOCKSIZE], &outlen);
+      cmac(key, input, AES_BLOCKSIZE, &buf[i*AES_BLOCKSIZE], &outlen, buf_capacity);
```
Comment (MrAnno) [inline modules/secure-logging/slog.c:501]:
```diff
- cmac(key, input, AES_BLOCKSIZE, &buf[i*AES_BLOCKSIZE], &outlen, buf_capacity);
+ cmac(key, input, AES_BLOCKSIZE, buf + (i*AES_BLOCKSIZE), &outlen, G_N_ELEMENTS(buf) - (i*AES_BLOCKSIZE));
```

File: modules/afsql/afsql.c [hunk 7]
```diff
@@ -384,15 +384,15 @@ afsql_dd_create_index(AFSqlDestDriver *self, const gchar *table, const gchar *co
...
       if ((strlen(table) + strlen(column)) > 25)
         {
 
-          guchar hash[MD5_DIGEST_LENGTH];
+          guchar hash[SHA_DIGEST_LENGTH];
           gchar hash_str[31];
           gchar *cat = g_strjoin("_", table, column, NULL);
 
-          MD5((guchar *)cat, strlen(cat), hash);
+          SHA1((guchar *)cat, strlen(cat), hash);
```
Comment (MrAnno) [inline modules/afsql/afsql.c:395]: We've discussed this IRL and we decided to keep the old MD5 index to be fully compatible with older versions.

A non-deprecated API can be used: EVP_Digest.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — DISMISSED** [review]

File: modules/cryptofuncs/cryptofuncs.c [hunk 8]
```diff
@@ -94,7 +94,12 @@ tf_hash_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent, gint
       return FALSE;
     }
   state->length = length;
-  md = EVP_get_digestbyname(strcmp(argv[0], "hash") == 0 ? "sha256" : argv[0]);
+  const char *digest_name = strcmp(argv[0], "hash") == 0 ? "sha256" : argv[0];
+  if (strcmp(digest_name, "md4") == 0)
+    {
+      msg_warning_once("WARNING: MD4 hash template function is deprecated and will be removed in future versions");
```
Comment (gaborznagy) [inline modules/cryptofuncs/cryptofuncs.c:100]: Sorry, I didn't review the whole branch, I just have this one comment:

I think we should send a deprecation warning if syslog-ng compiled with Openssl 3 too (and nost just in case with Openssl 1.1):
We could add a dummy template function (.e.g a TEMPLATE_FUNCTION_SIMPLE for md4 in case of Openssl 3.
There is also an example how to do error reporting in case of deprecation:

https://github.com/syslog-ng/syslog-ng/blob/f69ec0a98b34f43820df3157a95d8f52b6b50c63/modules/cryptofuncs/cryptofuncs.c#L97-L102

Thanks @MrAnno to confirm my idea.

**Review by MrAnno — APPROVED** [review]

**Review by alltilla — APPROVED** [review]

File: modules/cryptofuncs/cryptofuncs.c (hunk 8 above)
Comment (gaborznagy) [reply modules/cryptofuncs/cryptofuncs.c]: I've checked the updated version, looks good to me, thanks!

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]

**Review by alltilla — APPROVED** [review]

**Comment by MrAnno** [comment]
Please do not merge until @OverOrion verifies the latest version of this PR with OpenSSL 3.0.
(We currently don't have a CI job for it, so it is going to be done manually.)
